ledger/src/check_next_block.rs

```diff
@@ -129,6 +132,40 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         Ok(())
     }
 
+    /// Check that leaves in the subdag point to batches in other blocks that are valid.
+    fn check_block_subdag_leaves(&self, block: &Block<N>) -> Result<()> {
+        // Check if the block has a subdag.
+        let Authority::Quorum(subdag) = block.authority() else {
+            return Ok(());
+        };
+
+        // Store the IDs of all certificates in this subDAG.
+        // This allows determining which edges point to other subDAGs/blocks.
+        let subdag_certs: HashSet<_> = subdag.certificate_ids().collect();
+        let leaf_edges: HashSet<_> = subdag
+            .certificates()
+            .flat_map(|cert| cert.previous_certificate_ids().iter().map(|id| (cert.round() - 1, id)))
+            .filter(|(_, id)| !subdag_certs.contains(id))
+            .collect();
+
+        for (round, cert_id) in leaf_edges {
+            // Find the block for this certificate.
+            // We might check the same block multiple times, but, so far, the check
+            // simply ensures that the block is a valid ancestor.
+            match self.vm.block_store().get_block_for_certificate(cert_id, Some(round))? {
+                Some(prev_block) => {
+                    ensure!(
+                        prev_block.height() < block.height(),
+                        "Leaf is pointing to a block that is not an ancestor"
+                    );
+                }
+                None => bail!("Leaf is pointing to certificate that is not associated with a previous block"),
+            }
+        }
+
```

Comment:
There is a rare but possible case in which we need to relax these checks: if the sub-DAG spans the full 100 Subdag::MAX_ROUNDS, it is possible that the certificates in the smallest round of the sub-DAG have dangling references to previous certificates, because when a sub-DAG is committed, at most 100 rounds starting from the anchor (and going backwards) are considered, as it has emerged from [recent discussion](https://github.com/ProvableHQ/snarkOS/pull/3564/files).

Reply:
This should be addressed with the most recent change. I also added a test for it.